logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/handler/DeleteMessageHandler.kt

```diff
@@ -0,0 +1,91 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.logic.sync.receiver.handler
+
+import com.wire.kalium.logic.data.asset.AssetRepository
+import com.wire.kalium.logic.data.id.ConversationId
+import com.wire.kalium.logic.data.message.Message
+import com.wire.kalium.logic.data.message.MessageContent
+import com.wire.kalium.logic.data.message.MessageRepository
+import com.wire.kalium.logic.data.user.UserId
+import com.wire.kalium.logic.functional.onSuccess
+
+internal interface DeleteMessageHandler {
+    suspend operator fun invoke(
+        content: MessageContent.DeleteMessage,
+        conversationId: ConversationId,
+        senderUserId: UserId
+    )
+}
+
+internal class DeleteMessageHandlerImpl internal constructor(
+    private val messageRepository: MessageRepository,
+    private val assetRepository: AssetRepository,
+    private val selfUserId: UserId
+) : DeleteMessageHandler {
+    override suspend fun invoke(
+        content: MessageContent.DeleteMessage,
+        conversationId: ConversationId,
+        senderUserId: UserId
+    ) {
+        messageRepository.getMessageById(conversationId, content.messageId).onSuccess { messageToRemove ->
+            val isSelfSender = messageToRemove.senderUserId == selfUserId
+            val isOriginalEphemeral = (messageToRemove as? Message.Regular)?.expirationData != null
+            if (isSelfSender && isOriginalEphemeral) {
+                /*
+                 * if self is the sender and the message is ephemeral, delete the message
+                 * without verifying the sender of the delete message
+                 */
+                messageRepository.deleteMessage(
+                    messageUuid = messageToRemove.id,
+                    conversationId = messageToRemove.conversationId
+                )
+            } else {
+                val canBeDeleted = isSenderVerified(messageToRemove, senderUserId)
+                if (!canBeDeleted) return
+
+                if (isOriginalEphemeral) {
+                    messageRepository.deleteMessage(
+                        messageUuid = messageToRemove.id,
+                        conversationId = messageToRemove.conversationId
+                    )
+                } else {
+                    messageRepository.markMessageAsDeleted(
+                        messageUuid = messageToRemove.id,
+                        conversationId = messageToRemove.conversationId
+                    )
+                }
+            }
+            removeAssetIfExists(messageToRemove)
+        }
+    }
+
+    /**
+     * checks if the sender of the delete message is the same as the sender of the message to be deleted
+     */
+    private fun isSenderVerified(
+        message: Message,
+        deleteMessageSenderId: UserId
+    ): Boolean = deleteMessageSenderId == message.senderUserId
```

Comment:
if self is not the sender of the original message, we still need to verify that the sender of original message we still need to verify that the user asking for delete is the same as the original sender, or am i missing the point here

the flow if like this

1. when original message is Ephemeral and self is the sender -> then delete without checking the delete message sender.
2. otherwise we only delete if the sender of the delete message is the same as the original message.

removing this and only adding `if (!isSelfSender) return` means any conversation member can request delete to any message

Reply:
You're right